### DISCUSSION TITLE
Top level PR comment thread

### DISCUSSION THREAD

**Discussion by nickofthyme** [community/community#8469] [Pull Requests]
I wish there was a way to create a thread on top-level pull request review comments.

Often times I find I want to add a general comment that is not related to any specific line of code. For example, like https://github.com/elastic/elastic-charts/pull/1475#pullrequestreview-802918615. I have left comments on specific lines of code to be addressed but then I have a general comment on top of those.

The issue is, if this top-level comment is actionable such that I am requesting the author to change some code in a general sense or it might just be a general question that I would like a response to, there is not a great way to respond to this comment.

Currently, the best approach is something like https://github.com/elastic/elastic-charts/pull/1475#issuecomment-965649909, where we use so called _**Quote reply**_  via block-quotes (i.e.`>`) to address each part of the top-level comment in a new top-level comment.

<img width="924" alt="image" src="https://user-images.githubusercontent.com/19007109/220399081-483dbb61-2049-4048-b966-80f0460adbae.png">

The issue with this is that it creates a disconnect between the original comment and the response, as well as creating noise on the pr with several top-level comments. This issue is not _as_ important with issues (non-prs) because the top-level is often a back and forth response with a single idea where there is not often tangential ideas/threads active at the same time, though could be useful in some cases.

An ideal solution would be to enable a thread for top-level pr review comments. Maybe by default it is as it is currently, but with an option to create a thread off of the comment.

**Comment by christothes** [comment]
Enabling threading on top level comments by default would greatly improve to PR commenting experience.

**Comment by usernamenumber** [comment]
Came looking for someone else saying this, and I'm sad to see the lack of traction it's gotten. This is absolutely my single biggest issue with using GitHub. 

I've taken to replying to top-level comments on PRs by @ing the person on slack, and attaching my own "top-level" comments to the first line of code. The alternative is to end up filling the PR with unresolvable top-level messages that could represent any number of discussions all mixed in with one another. It's a huge mess, and the excellent way comment threads in the actual code are handled shows that it doesn't need to be like that.

**Comment by geoguide** [comment]
Please add this feature, it's very challenging when people leave questions or thoughts that need a response in top level comments.

**Comment by hivokas** [comment]
This is wonderful idea. Would be happy to see it!

**Comment by hey-robin** [comment]
It would be very useful to have discussion threads on top level comments just like we have on file diffs.

**Comment by p0fi** [comment]
I was working with GitLab very long in my previous position. Now moved to GitHub and im somehow shocked that this does not exist. Replying to a comment in a thread style is basically needed everywhere not just in top level RP comments. Its needed in issue comments as well. 

Please @github-staff have a look! 👀

**Comment by mochadwi** [comment]
awesome, they're working really hard on it! https://github.com/github/roadmap/issues/552

**Reply by SMR39** [reply]
this also applies to issues or just PRs ?

**Reply by bangank36** [reply]
2022 is about to end now, not sure if the feature can be shipped soon?

**Reply by richard-rance** [reply]
It looks like this was bumped to Q2 2023.

**Reply by Linths** [reply]
It seems it's been [moved to "Future"](https://github.com/orgs/github/projects/4247/views/1?pane=issue&itemId=8861247) now, so probably low priority. Too bad though, this feature would be great!

**Reply by hyperTwitch** [reply]
FYI that roadmap item is now closed as not planned. "Replying to PR comments is currently in an exploratory phase, so we have removed it from this roadmap while we develop more definition and certainty on our scope and timelines."

**Reply by fkunkler-bee** [reply]
@hyperTwitch Could you please give us an update on that topic?

**Comment by ColbyODonnell** [comment]
Azure DevOps has this feature.  Microsoft, please port it to GitHub.

**Comment by jdechicchis** [comment]
Please add this feature. It really helps to make the PR comment view cleaner. It would also be nice for the "Require conversation resolution before merging" to (optionally) apply to top-level comments

**Comment by eschapira-cs** [comment]
Makes a lot of sense. Often need to provide more general comments, or alternatives that could be considered.

**Comment by dpirkle-naut** [comment]
Please add this "feature" (seems more like a bug fix to me though).

**Comment by abrahamhunt** [comment]
So happy to see this on the road map. In addition to the problematic ability to follow the conversations of top level comments the tone of `Quoting` option has always felt weirdly aggressive to me which is not something I ever want to convey. 🙏 we get this as soon as we possibly can.

**Comment by bautrey37** [comment]
I see that this issue was moved from Q2 2023 to the Future. What is the reasoning for this?

**Comment by rakusvn-ntdat** [comment]
Please help to support this feature. 

In my current project, many people can put comments on the PR. It is too difficult to track all the comments and reply to them.

**Comment by rakusvn-vmvuong** [comment]
This feature available in gitlab, it so great, so please support add this feature to github.

**Comment by phoenixeliot** [comment]
We sometimes work around this by using file comments; slightly better than using line comments since it doesn't mislead the reader into thinking it's about a specific line of code.

That workaround is super awkward though and I wish we didn't have to use it.

**Comment by pierresisson** [comment]
This will be an incredible feature, it exists in the discussions but we really need it in the pull requests!

thanks github team :)

**Comment by castagninojose** [comment]
This is such a PITA for millons of developers on a daily basis. Why on earth is this not a thing yet?
Code review in github is a complete PAIN. Please, for the love of god, million of developers around the world NEED THIS.

**Comment by emplumb** [comment]
coming from gitlab (which I'm used to having issues) I'm shocked it's had this feature for years whereas github still does not

**Comment by hsorsky** [comment]
+1 on this feature request! 

there are often more high level comments I want to make on PRs that dont relate to any specific piece of code. file level comments helps, but this would have been preferable

**Comment by mpl1018** [comment]
Hi everyone! We've created a free GitHub App + Action that enables threads on PR issue comments and can also be fully integrated with Slack. You can [test it out in this PR](https://github.com/pullpo-io/cli/pull/7).
More info on how to use it [here](https://docs.pullpo.io/channels-full-sync). 

https://github.com/community/community/assets/44880660/5f77c50c-46f4-4de0-bdc7-c521af5bc882

@nickofthyme @christothes @usernamenumber @geoguide @hivokas @robnit @p0fi @mochadwi @SMR39 @bangank36 @Linths @richard-rance @ColbyODonnell @jdechicchis @eschapira-cs @dpirkle-naut @abrahamhunt @bautrey37 @rakusvn-ntdat @rakusvn-vmvuong @phoenixeliot @pierresisson @castagninojose @emplumb @hsorsky

**Reply by 0xdevalias** [reply]
I did a bit of an exploration into this feature and how it works in the comment thread on another discussion:

- https://github.com/orgs/community/discussions/5633#discussioncomment-12088921

The tl;dr seems to be that it likely requires a paid subscription for the app to use it.

But I also hacked together a basic Proof of Concept GitHub Action (largely untested) that should be able to achieve similar in the same basic way, but for free (though I still don't like how it pollutes PR's with unrelated files just as a placeholder to leave file-based comments):

> Based on the description of how this feature works, at a high level, it seems like the Pullpo app will just notice when someone has made a comment on the main PR, convert that to a new [file comment](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#:~:text=To%20comment%20directly%20on%20a%20file) on the `一pullpo一/pr_<PR_NUMBER>_threads.txt` placeholder file ([creating the comment as the user](https://github.com/pullpo-io/cli/pull/7#pullrequestreview-1948496305) if it has permission to; and [falling back to posting as the app bot](https://github.com/pullpo-io/cli/pull/7#pullrequestreview-1948895541) if it doesn't have permissions to post as that user), and then deletes the old top-level comment.
> 
> Based on that functionality.. you could probably implement something similar purely within a GitHub action alone if you really wanted to... though personally, I don't really like the idea of littering my PR's with an unrelated file just to serve as a placeholder for where to create these new comment threads on.
> 
> I haven't deeply reviewed/tested the following GitHub Action, but here's a quick proof-of-concept I knocked up for this using ChatGPT `o3-mini-high`. It can't post the new comment as the user themselves.. but aside from that I think it pretty much implements the same basic workflow/etc:
> 
> - Event: https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#issue_comment
> - Payload: https://docs.github.com/en/webhooks/webhook-events-and-payloads#issue_comment
> - Creating a comment: https://docs.github.com/en/rest/pulls/comments#create-a-review-comment-for-a-pull-request
> 
> ```yaml
> name: Convert PR Top-Level Comment to File Comment
> 
> on:
>   issue_comment:
>     types: [created]
> 
> env:
>   COMMITTER_NAME: "github-actions"
>   COMMITTER_EMAIL: "github-actions@github.com"
> 
> jobs:
>   convert-comment:
>     runs-on: ubuntu-latest
>     if: ${{ github.event.issue.pull_request != null }}
>     env:
>       PLACEHOLDER_FILE: "thread-comments/pr_${{ github.event.issue.number }}_threads.txt"
>     steps:
>       - name: Get PR details
>         id: pr_details
>         uses: actions/github-script@v6
>         with:
>           script: |
>             const pull_number = context.payload.issue.number;
>             const { data: pr } = await github.rest.pulls.get({
>               owner: context.repo.owner,
>               repo: context.repo.repo,
>               pull_number: pull_number
>             });
>             core.setOutput("head_ref", pr.head.ref);
>             core.setOutput("head_sha", pr.head.sha);
>           token: ${{ secrets.ADMIN_TOKEN }}
> 
>       - name: Checkout PR branch
>         uses: actions/checkout@v4
>         with:
>           ref: ${{ steps.pr_details.outputs.head_ref }}
>           token: ${{ secrets.ADMIN_TOKEN }}
>           fetch-depth: 0
> 
>       - name: Ensure placeholder file exists
>         id: ensure_placeholder
>         run: |
>           if [ -f "${PLACEHOLDER_FILE}" ]; then
>             echo "Placeholder file exists. Skipping creation."
>           else
>             mkdir -p "$(dirname "${PLACEHOLDER_FILE}")"
>             cat <<EOF > "${PLACEHOLDER_FILE}"
> # Threaded Discussion Placeholder
> 
> This file is used to anchor threaded comments for this pull request.
> Do not modify this file manually.
> 
> 
> EOF
>             git config user.name "${COMMITTER_NAME}"
>             git config user.email "${COMMITTER_EMAIL}"
>             git add "${PLACEHOLDER_FILE}"
>             git commit -m "Add placeholder file for threaded discussions"
>             git push origin ${{ steps.pr_details.outputs.head_ref }}
>           fi
> 
>       - name: Create file-level review comment on placeholder file
>         id: create_review_comment
>         uses: actions/github-script@v6
>         with:
>           script: |
>             const owner = context.repo.owner;
>             const repo = context.repo.repo;
>             const pull_number = context.payload.issue.number;
>             const commit_id = "${{ steps.pr_details.outputs.head_sha }}";
>             const placeholderPath = "${PLACEHOLDER_FILE}";
>             const body = `**Threaded comment by @${context.payload.comment.user.login}:**\n\n${context.payload.comment.body}`;
>             // Using subject_type "file" creates a file-level review comment.
>             const { data: reviewComment } = await github.rest.pulls.createReviewComment({
>               owner,
>               repo,
>               pull_number,
>               commit_id,
>               path: placeholderPath,
>               body: body,
>               subject_type: "file"
>             });
>             core.info("Created review comment with ID: " + reviewComment.id);
>           token: ${{ secrets.ADMIN_TOKEN }}
> 
>       - name: Delete original top-level comment
>         uses: actions/github-script@v6
>         with:
>           script: |
>             await github.rest.issues.deleteComment({
>               owner: context.repo.owner,
>               repo: context.repo.repo,
>               comment_id: context.payload.comment.id
>             });
>             core.info("Deleted original top-level comment");
>           token: ${{ secrets.ADMIN_TOKEN }}
> ```
> 
> _Originally posted by @0xdevalias in https://github.com/orgs/community/discussions/5633#discussioncomment-12088921_

---

**Edit:** Captured in this gist for posterity:

- https://gist.github.com/0xdevalias/7807fa0a8febe648a5e6f70a3a4ac0b6#convert-top-level-pr-comments-to-file-attached-review-comments-via-github-actions

**Comment by Pastromhaug** [comment]
Agreed with the others here that this is probably the simplest single improvement github could make to drastically improve the user experience for millions of developers. Feels like the stone ages quoting each other's previous comments and trying to piece together threads when multiple people are reviewing a PR.

**Comment by jgreffe** [comment]
Ouch https://github.com/github/roadmap/issues/552#issuecomment-2489160191

**Reply by TheGiraffe3** [reply]
That's sad.

**Reply by hey-robin** [reply]
ugh whyyyy! It's such a basic feature

**Reply by geoguide** [reply]
I don't understand why this is so difficult to do.

**Comment by binhle113** [comment]
jui

**Comment by 0xdevalias** [comment]
See also:

- https://github.com/orgs/community/discussions/5633

**Comment by stephendeweyEQ** [comment]
Please do this, it would be a major improvement for sure! I miss this from GitLab ;)

**Comment by emilte** [comment]
It's absurd that this is not a feature :O

**Comment by rewgoes** [comment]
Up!

**Comment by chadxz** [comment]
Another aspect of this is the combination of “require conversation resolution” and “auto merge” features would benefit if conversations at the top level of a pull request were also able to be marked resolved. This would allow people to leave actionable feedback and ask that it be resolved before merging.

right now this requires you to go find a file to leave that comment on, even if there’s no specific file that it should be left on.

**Reply by NetherGranite** [reply]
My exact use case! Very annoying coming from Azure DevOps

**Comment by bencutting164-cloud** [comment]
Totally agree with this having threaded replies for top-level PR comments would make discussions much cleaner. It’s messy when general feedback or questions end up scattered across multiple comments. This feature would definitely improve collaboration.

**Comment by jimpanzeeee** [comment]
PR comment threads when

**Comment by CarlosRuiz-globalqss** [comment]
+1 - is there a way to vote for interesting feature requests?

**Comment by andrewhill-socotra** [comment]
This is one of the top 10 most requested pull request features, and one of the top 40 most upvoted GitHub discussions of all time. [It was added to the roadmap](https://github.com/github/roadmap/issues/552), and then removed so that GitHub could "develop more definition and certainty on [their] scope and timelines". Now, over a year later, that scope still hasn't been defined.

If scope and timelines are the problem, then I'd like to offer a suggestion: Do whatever is easiest. Port GitHub discussion threads or simply re-use the file-based pull request threads. Right now we can't have an organized conversation, so any kind of thread would be an improvement. I'm certain that GitHub agrees this is an important issue that can be solved, because they've already solved it for discussions and pull request files.

If there are still unanswered questions about the scope, look and feel, features, etc., then please jump into the discussion and ask questions. There are 600+ people here to help figure out an answer.

**Comment by mgssf** [comment]
Please do this. Can't believe it doesn't exist.

**Comment by HongtaoYang** [comment]
Is this implemented or planned yet? Coming over from gitlab and can't believe that github doesn't support threading

**Comment by robobario** [comment]
This would be a great feature, it's hellish if you get two or three threads of conversation going on at the top level of a PR with just block quotes to try and connect things together. It's common to have a discussion that isn't directly related to a specific modified file, like overall direction of the PR, omissions, etc.

It's also frustrating because they look so similar to the threaded conversations it _feels like_ there should be a reply box right there. I imagine it confuses users to see reply entries in some places and not others.

**Comment by micke232** [comment]
Let's make it happen before the apocolyps?

**Comment by sebastienboisvert** [comment]
Comments in PRs that aren't related to file(s) need to be threadable (just like comments on files) - currently one has to 'quote reply' to preserve some semblance of context in replies, which creates extra unnecessary noise and is difficult to follow in the disjointed list of comments that can happen in normal PR discussions.

Additionally, these types of comments need to be 'resolvable' (again, _just like comments on files!_) to help clean up resolved discussions.

P.S. I find it unbelievable this obviously needed feature still isn't implemented nearly 6 years after its initial request. 🙄

**Comment by geoguide** [comment]
Just tell claude to implement it and release it tomorrow this is insane.

**Reply by sebastienboisvert** [reply]
Feels like GH will sooner implement naming "Lake America" than fix this clear lack of functionality.

**Reply by geoguide** [reply]
if this is somehow a complicated change the system is completely screwed up.

**Comment by stephendewey-se** [comment]
Please do this, it's the worst limitation of GitHub.
